doc/keynames/new_api.py

```diff
@@ -0,0 +1,1113 @@
+#!/bin/python3
```

Comment:
@markus2330 @mpranj I added this as a demo of how I would imagine the Elektra key name API. Take a look, if you want to, but please wait with comments until our meeting. There are many parts of this that can be changed and adapted to better fit any requirements.

If you want to play with the API, I recommend using `./new_api.py repl`. There is also `./new_api.py demo` which among other things prints a `KeySet` (list of `Key`s actually) in various formats to show how storage files might look.

Reply:
Hehe, I definitely prefer the idea to reserve `®elektra` (if we need to reserve anything at all, see discussions about having reservation internal to storage plugins. If it is only reserved internally, we might also reserve `®meta` and `®value`?). Also the other things written below "Demo of new key name API" sound like a big improvement to me.

About the base name escaping of `#`: maybe our misunderstanding is about if we expect someone using `keySetBaseName` "that he/she knows what he/she is doing". Imho the escaping of e.g. `#` is unnecessary, why not simply allow construing arrays (also broken ones without the proper number of `_` if someone wants) using it as a low-level interface? Then `keySetBaseName` would not need to "escape" anything in the unescaped key name and we still have the strong guarantees it had. Only `keyAddName` does the proper escaping of arrays (i.e. adding `_`) and might reject things. Btw. I am actually fine with `keyAddName` rejecting `®`, as long as  users can bypass these restrictions with `keySetBaseName` (and then run into the danger that these things might get interpreted by plugins!). Obviously this all depends on if we want semantic in key names at all.

Additional Interfaces like _KeyHierarchy I welcome very much but please not before 1.0 as this is very hard to get right (I did not look into the details of your proposal) and there are very few applications which actually need something like this (e.g. whole of KDE only uses flat configurations and with a few exceptions they are happy with that). But after 1.0 I definitely want something like this! (In some scenarios, possibly also in the spec plugin, it is the **much** more sane interface to work with. E.g. yajl could be rewritten with 10% of the code having such an interface. Maybe we start by adding such an interface internal to spec, and then make it public later?)

My current "perfect Elektra" would be:

- no semantic in key names
- arrays in the KeySet are only a *representation for arrays* but not real arrays (as arrays is also a rather rare feature, e.g. KDE does not need it),  are simply numbered (without starting with `#`, 50 escaped with `_50`, but tagged with `array` meta-data). From the representation users can get real arrays via other interfaces like _KeyHierarchy or the high-level API.
- keyAddName is a safe interface (in the sense it does not allow to construct weird things usually unwanted like numbers that are not prefixed with `_`) to add something to the key name
- keyAddBaseName is low-level raw interface to add something directly to the unescaped key name which can be dangerous and make unwanted things (and I am fine also saying this in the API docu). It is mostly needed for storage plugins but not for applications.

Just a few thoughts, no need to answer, let us discuss it in the next meeting.

Reply:
Just a few further notes and clarifications, discussion will happen in the meeting...

> I definitely prefer the idea to reserve ®elektra

The nice thing about reserving `®` in UTF-8 is that the encoding is `\xc2\xae`. `\xae` is actually the codepoint for ® in ISO-8859-1 and Windows-1252 (`\xc2` is Â), so even if an editor doesn't default to UTF-8, there is a good chance that `®elektra` will be recognizable.

> why not simply allow construing arrays (also broken ones without the proper number of `_` if someone wants) [...] Only `keyAddName` does the proper escaping of arrays (i.e. adding `_`)

I'm not quite sure I understood your idea about constructing arrays. Also: It is already the case that only `keyAddName` adds `_`s and I wouldn't suggested anything else.

What I maybe didn't put into the code comments, but would've explained in the meeting:
The idea with this Python API was that there are names and there are indices. Names are for maps, indices for arrays. `add_base_name` **always** adds a name and `add_array_index` **always** adds an index. Basic functionality of the functions is different. Key name parts starting with a single `#` are **always** array indices. `add_base_name` always adds a second `#` if it detects a `#` to make sure the intention of "add a name" is fulfilled.

> Additional Interfaces like _KeyHierarchy [...] there are very few applications which actually need something like this

`KeyHierarchy` was intended to help storage plugins and not applications. I agree that only few applications could use it. The `lcdexec` menu in LCDproc is a candidate, but the fact that I managed to implement it without a `KeyHierarachy` shows it isn't really needed.

A possibility (especially if we leave the plugin API unstable for 1.0) would be to explicitly state that `KeyHierarchy` is not yet part of the stable API and only patch releases, but not minor version updates will be compatible.

> as arrays is also a rather rare feature [...] without starting with `#`

I think this very much depends on what config format an application uses. If the application uses a JSON config, I think it is much more likely to use arrays than with INI. The concept of arrays definitely comes up often (e.g. multiple accounts, list of servers, etc.), but often you can work around the need for actual arrays.

IIRC when I started this PR last year, we said the `#` is a nice visual indicator and makes it easier to spot array elements within long key names. But maybe idea of separating names and indices would benefit from removing the `#`. It's quite common in programming that identifiers cannot start with a digit. However, prefixing with `_` is a common way around that, but we use `_` for number. Let's discuss this in the meeting. 

> keyAddName is a safe interface [...] keyAddBaseName is low-level raw interface

Maybe we need another level. `keyAddName` works with the escaped name. That can be tedious at times.

In the Python API I thought about adding a `add_last_part` which just does `self.__parts.append (part)` without any checks or processing. I didn't like the idea. Without `add_last_part` you have to think about what you are doing. That makes it harder to accidentally create something broken. It also avoids some problems with aliasing/uniqueness of key names.

> [keyAddBaseName] is mostly needed for storage plugins but not for applications.

I think this depends on the application. For a static config with a fixed number of keys, yes `keyAddBaseName` isn't needed. If there are dynamic parts e.g. there is a  `/mail/[account_name]/...` subset for each account [1], then `keyAddBaseName` is indeed useful.

I would also say that `keyAddBaseName` is useful in _all_ plugins and libraries, not just storage plugins.